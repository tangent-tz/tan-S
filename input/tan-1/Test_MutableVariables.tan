main {
    # 1. declare a 'var' variable with simple integer, float:
    var i := 123;
    var f := 45.678;
    print 1\n i \n f \n\n;

    # Expected output:
    # 123
    # 45.678

    #================================================================================

    # 2. re-assignment of 'var' variable to new values:
    i := 56;
    f := 3546.00055;
    print 2\n i \n f \n\n;

    # Expected output:
    # 56
    # 3546.00055

    #================================================================================

    # 3. re-assignment of 'var' variable to some other 'const' variable of the same type:
    const varone := 99;
    i := varone;

    const vartwo := 01.245;
    f := vartwo;
    print 3\n i \n f \n\n;

    # Expected output:
    # 99
    # 1.245

    #================================================================================

    # 4. re-assignment of 'var' variable to some other 'var' variable of the same type:
    var varthree := 1;
    i := varthree;

    var varfour := 1.234;
    f := varfour;
    print 4\n i \n f \n\n;

    # Expected output:
    # 1
    # 1.234

    #================================================================================

    # 5. re-assignment of 'var' variable to some expression (with literals):
    i := 1 + 2 + 3;
    f := 1.1 + 2.22 + 3.333;
    print 5\n i \n f \n\n;

    # Expected output:
    # 6
    # 6.653

    #================================================================================

    # 6. re-assignment of 'var' variable to some expression (with variables):
    var varfive := 230;
    var varsix := 1000000;
    const varseven := 97;
    i := varfive + varsix + varseven;

    var vareight := 14354.004;
    var varnine := 3.60;
    const varten := 44.555;
    f := vareight + varnine + varten;

    print 6\n i \n f \n\n;

    # Expected output:
    # 1000327
    # 14402.159

    #================================================================================

    # 7. re-assignment of 'var' variable to some expression that references itself:
    var vareleven := 87;
    i := i + 100 + vareleven;

    var vartwelve := 9000000.12;
    f := f + 22.570 + vartwelve;

    print 7\n i \n f \n\n;

    # Expected output:
    # 1000514
    # 9014424.849
}