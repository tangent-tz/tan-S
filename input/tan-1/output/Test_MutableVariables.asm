        Jump         $$main                    
        DLabel       $eat-location-zero        
        DataZ        8                         
        DLabel       $print-format-integer     
        DataC        37                        %% "%d"
        DataC        100                       
        DataC        0                         
        DLabel       $print-format-boolean     
        DataC        37                        %% "%s"
        DataC        115                       
        DataC        0                         
        DLabel       $print-format-float       
        DataC        37                        %% "%f"
        DataC        102                       
        DataC        0                         
        DLabel       $print-format-character   
        DataC        37                        %% "%c"
        DataC        99                        
        DataC        0                         
        DLabel       $print-format-string      
        DataC        37                        %% "%s"
        DataC        115                       
        DataC        0                         
        DLabel       $print-format-newline     
        DataC        10                        %% "\n"
        DataC        0                         
        DLabel       $print-format-space       
        DataC        32                        %% " "
        DataC        0                         
        DLabel       $print-format-tab         
        DataC        9                         %% "\t"
        DataC        0                         
        DLabel       $boolean-true-string      
        DataC        116                       %% "true"
        DataC        114                       
        DataC        117                       
        DataC        101                       
        DataC        0                         
        DLabel       $boolean-false-string     
        DataC        102                       %% "false"
        DataC        97                        
        DataC        108                       
        DataC        115                       
        DataC        101                       
        DataC        0                         
        DLabel       $errors-general-message   
        DataC        82                        %% "Runtime error: %s\n"
        DataC        117                       
        DataC        110                       
        DataC        116                       
        DataC        105                       
        DataC        109                       
        DataC        101                       
        DataC        32                        
        DataC        101                       
        DataC        114                       
        DataC        114                       
        DataC        111                       
        DataC        114                       
        DataC        58                        
        DataC        32                        
        DataC        37                        
        DataC        115                       
        DataC        10                        
        DataC        0                         
        Label        $$general-runtime-error   
        PushD        $errors-general-message   
        Printf                                 
        Halt                                   
        DLabel       $errors-int-divide-by-zero 
        DataC        105                       %% "integer divide by zero"
        DataC        110                       
        DataC        116                       
        DataC        101                       
        DataC        103                       
        DataC        101                       
        DataC        114                       
        DataC        32                        
        DataC        100                       
        DataC        105                       
        DataC        118                       
        DataC        105                       
        DataC        100                       
        DataC        101                       
        DataC        32                        
        DataC        98                        
        DataC        121                       
        DataC        32                        
        DataC        122                       
        DataC        101                       
        DataC        114                       
        DataC        111                       
        DataC        0                         
        Label        $$i-divide-by-zero        
        PushD        $errors-int-divide-by-zero 
        Jump         $$general-runtime-error   
        DLabel       $errors-float-divide-by-zero 
        DataC        102                       %% "float divide by zero"
        DataC        108                       
        DataC        111                       
        DataC        97                        
        DataC        116                       
        DataC        32                        
        DataC        100                       
        DataC        105                       
        DataC        118                       
        DataC        105                       
        DataC        100                       
        DataC        101                       
        DataC        32                        
        DataC        98                        
        DataC        121                       
        DataC        32                        
        DataC        122                       
        DataC        101                       
        DataC        114                       
        DataC        111                       
        DataC        0                         
        Label        $$f-divide-by-zero        
        PushD        $errors-float-divide-by-zero 
        Jump         $$general-runtime-error   
        DLabel       $usable-memory-start      
        DLabel       $global-memory-block      
        DataZ        128                       
        Label        $$main                    
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushI        123                       
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushF        45.678000                 
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushI        56                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushF        3546.000550               
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        12                        
        Add                                    %% varone
        PushI        99                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushD        $global-memory-block      
        PushI        12                        
        Add                                    %% varone
        LoadI                                  
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        16                        
        Add                                    %% vartwo
        PushF        1.245000                  
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushD        $global-memory-block      
        PushI        16                        
        Add                                    %% vartwo
        LoadF                                  
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        24                        
        Add                                    %% varthree
        PushI        1                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushD        $global-memory-block      
        PushI        24                        
        Add                                    %% varthree
        LoadI                                  
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        28                        
        Add                                    %% varfour
        PushF        1.234000                  
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushD        $global-memory-block      
        PushI        28                        
        Add                                    %% varfour
        LoadF                                  
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushI        1                         
        PushI        2                         
        Add                                    
        PushI        3                         
        Add                                    
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushF        1.100000                  
        PushF        2.220000                  
        FAdd         null                      
        PushF        3.333000                  
        FAdd         null                      
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        36                        
        Add                                    %% varfive
        PushI        230                       
        StoreI                                 
        PushD        $global-memory-block      
        PushI        40                        
        Add                                    %% varsix
        PushI        1000000                   
        StoreI                                 
        PushD        $global-memory-block      
        PushI        44                        
        Add                                    %% varseven
        PushI        97                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushD        $global-memory-block      
        PushI        36                        
        Add                                    %% varfive
        LoadI                                  
        PushD        $global-memory-block      
        PushI        40                        
        Add                                    %% varsix
        LoadI                                  
        Add                                    
        PushD        $global-memory-block      
        PushI        44                        
        Add                                    %% varseven
        LoadI                                  
        Add                                    
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        48                        
        Add                                    %% vareight
        PushF        14354.004000              
        StoreF                                 
        PushD        $global-memory-block      
        PushI        56                        
        Add                                    %% varnine
        PushF        3.600000                  
        StoreF                                 
        PushD        $global-memory-block      
        PushI        64                        
        Add                                    %% varten
        PushF        44.555000                 
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushD        $global-memory-block      
        PushI        48                        
        Add                                    %% vareight
        LoadF                                  
        PushD        $global-memory-block      
        PushI        56                        
        Add                                    %% varnine
        LoadF                                  
        FAdd         null                      
        PushD        $global-memory-block      
        PushI        64                        
        Add                                    %% varten
        LoadF                                  
        FAdd         null                      
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        72                        
        Add                                    %% vareleven
        PushI        87                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushI        100                       
        Add                                    
        PushD        $global-memory-block      
        PushI        72                        
        Add                                    %% vareleven
        LoadI                                  
        Add                                    
        StoreI                                 
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% i
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        76                        
        Add                                    %% vartwelve
        PushF        9000000.120000            
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushF        22.570000                 
        FAdd         null                      
        PushD        $global-memory-block      
        PushI        76                        
        Add                                    %% vartwelve
        LoadF                                  
        FAdd         null                      
        StoreF                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% f
        LoadF                                  
        PushD        $print-format-float       
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        84                        
        Add                                    %% constvalue
        PushI        10                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        84                        
        Add                                    %% constvalue
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        88                        
        Add                                    %% varvalue
        PushI        5                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        88                        
        Add                                    %% varvalue
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        88                        
        Add                                    %% varvalue
        PushI        8                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        88                        
        Add                                    %% varvalue
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        92                        
        Add                                    %% consta
        PushI        2                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        96                        
        Add                                    %% constb
        PushI        3                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        100                       
        Add                                    %% result
        PushD        $global-memory-block      
        PushI        92                        
        Add                                    %% consta
        LoadI                                  
        PushD        $global-memory-block      
        PushI        96                        
        Add                                    %% constb
        LoadI                                  
        Add                                    
        StoreI                                 
        PushD        $global-memory-block      
        PushI        100                       
        Add                                    %% result
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        104                       
        Add                                    %% outervalue
        PushI        10                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        108                       
        Add                                    %% outervalue
        PushI        20                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        108                       
        Add                                    %% outervalue
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        104                       
        Add                                    %% outervalue
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        108                       
        Add                                    %% conststring
        DLabel       _string_1_                
        DataI        3                         
        DataI        9                         
        DataI        5                         
        DataC        72                        %% "Hello"
        DataC        101                       
        DataC        108                       
        DataC        108                       
        DataC        111                       
        DataC        0                         
        PushD        _string_1_                
        StoreI                                 
        PushD        $global-memory-block      
        PushI        112                       
        Add                                    %% varinteger
        PushI        10                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        108                       
        Add                                    %% conststring
        LoadI                                  
        PushI        12                        
        Add                                    
        PushD        $print-format-string      
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        112                       
        Add                                    %% varinteger
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        116                       
        Add                                    %% constx
        PushI        2                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        120                       
        Add                                    %% vary
        PushI        3                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        124                       
        Add                                    %% varz
        PushD        $global-memory-block      
        PushI        116                       
        Add                                    %% constx
        LoadI                                  
        PushD        $global-memory-block      
        PushI        120                       
        Add                                    %% vary
        LoadI                                  
        PushI        2                         
        Multiply                               
        Add                                    
        StoreI                                 
        PushD        $global-memory-block      
        PushI        124                       
        Add                                    %% varz
        LoadI                                  
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        Halt                                   
