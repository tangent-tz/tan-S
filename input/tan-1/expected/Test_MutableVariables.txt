123
45.678001
56
3546.000488
99
1.245000
1
1.234000
6
6.653000
1000327
14402.158906
1000514
9014424.728906
10
5
8
5
20
10
Hello
10
8

